server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java

```diff
@@ -0,0 +1,386 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.extensions;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.opensearch.Version;
+import org.opensearch.action.admin.cluster.node.info.PluginsAndModules;
+import org.opensearch.action.admin.cluster.state.ClusterStateResponse;
+import org.opensearch.cluster.ClusterSettingsResponse;
+import org.opensearch.cluster.LocalNodeResponse;
+import org.opensearch.cluster.node.DiscoveryNode;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.io.FileSystemUtils;
+import org.opensearch.common.io.stream.StreamInput;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.transport.TransportAddress;
+import org.opensearch.discovery.PluginRequest;
+import org.opensearch.discovery.PluginResponse;
+import org.opensearch.extensions.ExtensionsSettings.Extension;
+import org.opensearch.index.IndexModule;
+import org.opensearch.index.IndexService;
+import org.opensearch.index.IndicesModuleNameResponse;
+import org.opensearch.index.IndicesModuleRequest;
+import org.opensearch.index.IndicesModuleResponse;
+import org.opensearch.index.shard.IndexEventListener;
+import org.opensearch.indices.cluster.IndicesClusterStateService;
+import org.opensearch.node.ReportingService;
+import org.opensearch.plugins.PluginInfo;
+import org.opensearch.threadpool.ThreadPool;
+import org.opensearch.transport.TransportException;
+import org.opensearch.transport.TransportResponse;
+import org.opensearch.transport.TransportResponseHandler;
+import org.opensearch.transport.TransportService;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
+
+/**
+ * The main class for Plugin Extensibility
+ *
+ * @opensearch.internal
+ */
+public class ExtensionsOrchestrator implements ReportingService<PluginsAndModules> {
```

Comment:
We don't really have other orchestrators in the codebase. Why did we pick this over, say, manager?

Reply:
@owaiskazi19 any thoughts on this?  I wasn't working on extensions when the file was created.

Reply:
This file is equivalent to what Plugins architecture currently have as [PluginsService.java](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/plugins/PluginsService.java).

Reply:
+1 to ExtensionsManager